ktxspec.adoc

```diff
@@ -837,7 +897,178 @@ individual `keyAndValueByteLength` fields.
 ==== supercompressionGlobalData
 An array of data used by certain supercompression schemes that must
 be available before any mip level can be expanded. Must start on an
-the next 8-byte boundary following the key/value data.
+the next 8-byte boundary following the key/value data. For some schemes,
+the specification for this data block is given below.
+
+[#basisu_gd]
+==== Basis Universal Global Data
+
+Basis Universal combines encoding to ETC1S (a block-compressed
+texture format that is a subset of ETC1, see <<KDF13>>) then
+supercompressing the result using a lossless compression scheme
+with similarites to LZMA.  It creates a global codebook referenced
+by each supercompressed image. The global data block contains this
+codebook in the form of supercompressed ETC1S endpoint and selector
+data and the huffman tables that are core to lossless supercompression.
+
+It also contains an array of _image descriptors_ with flags for each image
+and the offset and length within the image's mip level of the data for
+that image. Images may consist of 1 or 2 _slices_ depending on the
+components of the pre-deflation image as detailed below.
+
+_One-component images_ must have the component value replicated in all 3
+components of the _rgb_ slice of the compressed data and have no _alpha_
```

Comment:
Given that slice data is generated by Basis Universal encoder from an unknown source data, it looks like the KTX2 spec technically disallows ambiguous source/target combinations. For example:
- One-component image can be encoded to BasisU with RGB channels populated with the same value. Transcoding it to BC1/ETC1: three RGB slice channels populating three RGB texture sampler channels with the same value.
- One-component image can be encoded to BasisU with only green channel populated with zeros in red and blue . Transcoding it to BC1/ETC1: three RGB slice channels populating three RGB texture sampler channels with data only in green.

This (and subsequent) restrictions are added to potentially reduce endpoints footprint and to ensure consistency of transcoded images when the runtime transcoded format is uncompressed or suboptimal (otherwise 1- and 2-component images could always use green channel), right?

Reply:
Yes I want to restrict the possibilities so as to make behavior consistent and predictable and to reduce endpoint footprint. I'm basically following the undocumented behaviour of the BasisU encoder. It works only on 4 component images and uses lodepng to load png files. When converting to RGBA lodepng replicates "grey" to R, G & B for 1 and 2 component images. The BasisU encoder uses the RGB values unchanged. `libktx` does similar when preparing in memory RGBA images for the encoder. The encoder copies the alpha value to all 3 channels of the alpha slice.

Note that the transcoder does not currently have an option to transcode to uncompressed 1 or 2 component images only 3 or 4 component.

Reply:
It may be these particular restrictions, based on behaviour inherited by the BasisU encoder from lodepng, are not the right restrictions. In PNG, 1 & 2 component images are labelled grey and grey-alpha so swizzling the "grey" component to r, g & b makes sense. In KTX, OpenGL & Vulkan 1 & 2 component images are known as R & RG. I think the restrictions outlined make sense in this case too but I am open to alternatives.

Another issue is that currently the encoder supports a `separateRGToRGB_A` option. This causes the R & G components of an RGB{,A} input image to be swizzled to RGB and A respectively.  This option is copied from BasisU. I am not sure there is any utility to it. Is there any content creation pipeline that creates tangent space normal maps in 3 or 4 component textures? Those being the the advertised use of the option. [Note: if you were to pass a 2 component .png file to the BasisU encoder and set this option, the result would be wrong. The "grey" component would end up in all 3 components of the RGB *and* alpha slices.]

Reply:
>  In PNG, 1 & 2 component images are labelled grey and grey-alpha so swizzling the "grey" component to r, g & b makes sense. In KTX, OpenGL & Vulkan 1 & 2 component images are known as R & RG.

Legacy versions of OpenGL had Luminance and Luminance-Alpha (1- and 2-component) formats that semantically match PNG. I think that the current approach is correct.

> This causes the R & G components of an RGB{,A} input image to be swizzled to RGB and A. This option is copied from BasisU. I am not sure there is any utility to it. 

Original 2-component image may be authored in JPEG which supports only 1 or 3-component images. Also, it could be an RG texture saved in PNG verbatim (with zeros in blue channel). This encoder option gives some additional flexibility, but maybe the right approach is to generalize input data swizzling so this option would translate to something like `--swizzle-inputs rrrg`.

Reply:
Okay. Implementation of the option is already by a generalized swizzle. I think exposing that is a good idea. But that is a matter for `libktx` and `toktx` not the spec. Specwise I'll leave this as is but add some words so offering a swizzle option doesn't conflict.